---
title: "SenPred"
author: "Bethany Hughes"
date: "13 September 2023"
output:
  rmarkdown::pdf_document:
    toc: TRUE
urlcolor: blue
---
Machine learning model built from single cell RNA sequencing of replicative senescent and early proliferative human detmal fibroblasts

Funding sources
- BBSRC iCASE Unilever PhD studentship (BB/V509668/1)

### Table of Contents
* [Lead contacts] 
* [Project background] 
* [Input data] 
* [Installation]
* [File overview]
* [Analysis pipeline]
* [Output] 


### Lead contacts
**Primary author**: Beth Hughes, Queen Mary University of London, [b.k.hughes\@qmul.ac.uk](mailto:b.k.hughes@qmul.ac.uk)  
**Principal Investigator**: Professor Cleo Bishop, Queen Mary University of London,  [c.l.bishop\@qmul.ac.uk](mailto:c.l.bishop@qmul.ac.uk)  

### Project background
This analysis was run based on the Seurat 4.3.0.1 and scPred 1.9.2 vignettes. This analysis has been version controlled on GitHub, and can be accessed via the following URL <https://github.com/bethk-h/SenPred_HDF>. 

The input data was generated by Dr Deborah Milligan, and sequenced using the v2 Chromium single cell 3' Kit (10X Genomics), and Illumina NextSeq 500 High Output Run Sequencing platform using paired end sequencing.  

### Input data

The input data for this project can be found on GEO, accession number ?????. The contents of the Inputfiles folder will not be tracked via this repository, and will need to be downloaded before running. These downloaded files need to be placed into a folder called Inputfiles within the R project.


### Installation
R version 4.3.1 (2023-06-16 ucrt)

Package versions are stored in [renv.lock](renv.lock) file. To use this folder, one must remove the line "renv.lock" from the [.gitignore](.gitignore) file. Relevant package versions can then be installed using renv::restore(). 


### File overview
##### R Scripts
All scripts can be found in the [R_scripts](R_scripts) folder. The number prefix indicates the order the files should be run (from 1 to 12). To run the complete analysis, first run [95_Make_Clean.rmd]() to create or clean the results folder. Then, run [99_Run_All.rmd]() to run the scripts in order.

### Analysis pipeline
* [1_HDF_data.Rmd](R_scripts/1_HDF_data.Rmd) - loading data and converting into Seurat object.
* [2_HDF_filtering_normalisation.Rmd](R_scripts/2_HDF_filtering_normalisation.Rmd) - Filtering and normalisation of the Seurat object.
* [3_HDF_2DCells.Rmd](R_scripts/3_HDF_2DCells.Rmd) - filtering Seurat object to select the 2D cells, and investigating clustering and differential expression of 2D fibroblasts.
* [4_Scpred_2D.Rmd](R_scripts/4_Scpred_2D.Rmd) - Building machine learning models of proliferative and senescent HDFs in 2D, using ScPred package (see https://github.com/powellgenomicslab/scPred).
* [5_ScPred2d_Chan.Rmd](R_scripts/5_ScPred2d_Chan.Rmd) - Testing ML models on external replicative senescence fibroblast single cell RNA seq data set from Chan et al. (Chan, M., Yuan, H., Soifer, I., Maile, T.M., Wang, R.Y., Ireland, A., O'Brien, J.J., Goudeau, J., Chan, L.J., Vijay, T. and Freund, A., 2022. Novel insights from a multiomics dissection of the Hayflick limit. Elife, 11, p.e70283.).
* [6_ScPred2D_Chan_EPESDS.Rmd](R_scripts/6_ScPred2D_Chan_EPESDS.Rmd) - Building PDL50 cells from Chan et al. into the model, as ES or 'Early Senescent' cells.
* [7_Tabib.Rmd](R_scripts/7_Tabib.Rmd) - Testing the 2D model on external whole skin in vivo single cell RNA seq dataset from Tabib et al. (Tabib, T., Morse, C., Wang, T., Chen, W. and Lafyatis, R., 2018. SFRP2/DPP4 and FMO1/LSP1 define major fibroblast populations in human skin. Journal of Investigative Dermatology, 138(4), pp.802-810.)
* [8_soleboldo.Rmd](R_scripts/8_soleboldo.Rmd) - Testing the 2D model on external whole skin in vivo single cell RNA seq dataset from Solé-Boldo et al. (Solé-Boldo, L., Raddatz, G., Schütz, S., Mallm, J.P., Rippe, K., Lonsdorf, A.S., Rodríguez-Paredes, M. and Lyko, F., 2020. Single-cell transcriptomes of the human skin reveal age-related loss of fibroblast priming. Communications biology, 3(1), p.188.)
* [9_ScPred_3D.Rmd](R_scripts/9_ScPred_3D.rmd) - Building an ML model to predict senescence of 3D fibroblasts.
* [10_3Dpred_Tabib.Rmd](R_scripts/10_3Dpred_Tabib.Rmd) - Testing the 3D model on external whole skin in vivo single cell RNA seq dataset from Tabib et al. (Tabib, T., Morse, C., Wang, T., Chen, W. and Lafyatis, R., 2018. SFRP2/DPP4 and FMO1/LSP1 define major fibroblast populations in human skin. Journal of Investigative Dermatology, 138(4), pp.802-810.)
* [11_3Dpred_SoleBoldo.Rmd](R_scripts/11_3Dpred_SoleBoldo.Rmd) - Testing the 3D model on external whole skin in vivo single cell RNA seq dataset from Solé-Boldo et al. (Solé-Boldo, L., Raddatz, G., Schütz, S., Mallm, J.P., Rippe, K., Lonsdorf, A.S., Rodríguez-Paredes, M. and Lyko, F., 2020. Single-cell transcriptomes of the human skin reveal age-related loss of fibroblast priming. Communications biology, 3(1), p.188.)


### Output
The resulting output files are saved to the [Output](Output) folder in this repository. This will not be tracked in GitHub. 

Please note, each time the script is run, the partitioning of training and testing dataset is random and therefore slightly different. If you want to run the scripts multiple times with reproducible outputs, you must copy the training and testing dataset files from [Output/Seurat_Objects](Output/Seurat_Objects), and save them to Input files. You must then remove the # in front of "training <- ReadRDS..." and "testing <- ReadRDS" in the scripts [4_Scpred_2D.Rmd](R_scripts/4_Scpred_2D.Rmd), [6_ScPred2D_Chan_EPESDS.Rmd](R_scripts/6_ScPred2D_Chan_EPESDS.Rmd), and [9_ScPred_3D.Rmd](R_scripts/9_ScPred_3D.rmd).
 

##### Session information


```r
sessionInfo()
#> R version 4.3.1 (2023-06-16 ucrt)
#> Platform: x86_64-w64-mingw32/x64 (64-bit)
#> Running under: Windows 10 x64 (build 19045)
#> 
#> Matrix products: default
#> 
#> 
#> locale:
#> [1] LC_COLLATE=English_United States.utf8  LC_CTYPE=English_United States.utf8   
#> [3] LC_MONETARY=English_United States.utf8 LC_NUMERIC=C                          
#> [5] LC_TIME=English_United States.utf8    
#> 
#> time zone: America/New_York
#> tzcode source: internal
#> 
#> attached base packages:
#> [1] stats4    stats     graphics  grDevices datasets  utils     methods   base     
#> 
#> other attached packages:
#>  [1] ggpubr_0.6.0                png_0.1-8                   MLmetrics_1.1.1            
#>  [4] factoextra_1.0.7            janitor_2.2.0               caret_6.0-94               
#>  [7] lattice_0.21-8              magrittr_2.0.3              scPred_1.9.2               
#> [10] reshape2_1.4.4              sctransform_0.3.5           cowplot_1.1.1              
#> [13] RCurl_1.98-1.12             lubridate_1.9.2             forcats_1.0.0              
#> [16] stringr_1.5.0               purrr_1.0.1                 readr_2.1.4                
#> [19] tidyr_1.3.0                 tibble_3.2.1                tidyverse_2.0.0            
#> [22] ggplot2_3.4.2               tidyseurat_0.6.1            ttservice_0.3.6            
#> [25] readxl_1.4.3                patchwork_1.1.2             SeuratObject_4.1.3         
#> [28] Seurat_4.3.0.1              dplyr_1.1.2                 rmarkdown_2.23             
#> [31] here_1.0.1                  SeuratWrappers_0.3.1        monocle3_1.3.4             
#> [34] SingleCellExperiment_1.22.0 SummarizedExperiment_1.30.2 GenomicRanges_1.52.0       
#> [37] GenomeInfoDb_1.36.3         IRanges_2.34.1              S4Vectors_0.38.1           
#> [40] MatrixGenerics_1.12.3       matrixStats_1.0.0           Biobase_2.60.0             
#> [43] BiocGenerics_0.46.0        
#> 
#> loaded via a namespace (and not attached):
#>   [1] spatstat.sparse_3.0-2   bitops_1.0-7            httr_1.4.6              RColorBrewer_1.1-3     
#>   [5] backports_1.4.1         tools_4.3.1             utf8_1.2.3              R6_2.5.1               
#>   [9] lazyeval_0.2.2          uwot_0.1.16             mgcv_1.9-0              withr_2.5.0            
#>  [13] sp_2.0-0                gridExtra_2.3           progressr_0.13.0        cli_3.6.1              
#>  [17] textshaping_0.3.6       spatstat.explore_3.2-1  labeling_0.4.2          sass_0.4.7             
#>  [21] spatstat.data_3.0-1     proxy_0.4-27            ggridges_0.5.4          pbapply_1.7-2          
#>  [25] systemfonts_1.0.4       R.utils_2.12.2          harmony_0.1.1           parallelly_1.36.0      
#>  [29] limma_3.56.2            rstudioapi_0.15.0       generics_0.1.3          ica_1.0-3              
#>  [33] spatstat.random_3.1-5   car_3.1-2               Matrix_1.6-0            ggbeeswarm_0.7.2       
#>  [37] fansi_1.0.4             abind_1.4-5             R.methodsS3_1.8.2       terra_1.7-46           
#>  [41] lifecycle_1.0.3         yaml_2.3.7              carData_3.0-5           snakecase_0.11.0       
#>  [45] recipes_1.0.6           Rtsne_0.16              grid_4.3.1              promises_1.2.0.1       
#>  [49] crayon_1.5.2            miniUI_0.1.1.1          pillar_1.9.0            knitr_1.43             
#>  [53] boot_1.3-28.1           future.apply_1.11.0     codetools_0.2-19        leiden_0.4.3           
#>  [57] glue_1.6.2              data.table_1.14.8       remotes_2.4.2.1         vctrs_0.6.3            
#>  [61] cellranger_1.1.0        gtable_0.3.3            assertthat_0.2.1        kernlab_0.9-32         
#>  [65] cachem_1.0.8            gower_1.0.1             xfun_0.39               S4Arrays_1.0.6         
#>  [69] mime_0.12               prodlim_2023.03.31      survival_3.5-5          timeDate_4022.108      
#>  [73] iterators_1.0.14        hardhat_1.3.0           lava_1.7.2.1            ellipsis_0.3.2         
#>  [77] fitdistrplus_1.1-11     ROCR_1.0-11             ipred_0.9-14            nlme_3.1-162           
#>  [81] RcppAnnoy_0.0.21        rprojroot_2.0.3         bslib_0.5.0             irlba_2.3.5.1          
#>  [85] vipor_0.4.5             KernSmooth_2.23-22      rpart_4.1.19            colorspace_2.1-0       
#>  [89] nnet_7.3-19             ggrastr_1.0.2           tidyselect_1.2.0        compiler_4.3.1         
#>  [93] mda_0.5-4               DelayedArray_0.26.7     plotly_4.10.2           scales_1.2.1           
#>  [97] lmtest_0.9-40           digest_0.6.33           goftest_1.2-3           spatstat.utils_3.0-3   
#> [101] minqa_1.2.5             XVector_0.40.0          htmltools_0.5.5         pkgconfig_2.0.3        
#> [105] lme4_1.1-34             highr_0.10              fastmap_1.1.1           rlang_1.1.1            
#> [109] htmlwidgets_1.6.2       shiny_1.7.4.1           farver_2.1.1            jquerylib_0.1.4        
#> [113] zoo_1.8-12              jsonlite_1.8.7          ModelMetrics_1.2.2.2    R.oo_1.25.0            
#> [117] GenomeInfoDbData_1.2.10 munsell_0.5.0           Rcpp_1.0.11             viridis_0.6.4          
#> [121] reticulate_1.30         stringi_1.7.12          pROC_1.18.4             zlibbioc_1.46.0        
#> [125] MASS_7.3-60             plyr_1.8.8              parallel_4.3.1          listenv_0.9.0          
#> [129] ggrepel_0.9.3           deldir_1.0-9            splines_4.3.1           tensor_1.5             
#> [133] hms_1.1.3               igraph_1.5.0.1          spatstat.geom_3.2-4     ggsignif_0.6.4         
#> [137] evaluate_0.21           renv_1.0.0              BiocManager_1.30.21.1   nloptr_2.0.3           
#> [141] tzdb_0.4.0              foreach_1.5.2           httpuv_1.6.11           RANN_2.6.1             
#> [145] polyclip_1.10-4         future_1.33.0           scattermore_1.2         rsvd_1.0.5             
#> [149] broom_1.0.5             xtable_1.8-4            e1071_1.7-13            rstatix_0.7.2          
#> [153] later_1.3.1             viridisLite_0.4.2       class_7.3-22            ragg_1.2.5             
#> [157] beeswarm_0.4.0          cluster_2.1.4           timechange_0.2.0        globals_0.16.2
```

This document was processed on: 2023-09-13.


