---
title: "16_Ganier"
author: "BH"
date: "2024-03-21"
output:
  prettydoc::html_pretty:
    theme: cayman
    toc: yes
---
---
title: "Ganier"
author: "BH"
date: "2024-01-15"
output: html_document
---

# Adding an additional whole skin dataset from Ganier et al. This data was downloaded from ArrayExpress under “E- MTAB- 13085”, from the following paper:
Ganier, C., Mazin, P., Herrera-Oropeza, G., Du-Harpur, X., Blakeley, M., Gabriel, J., Predeus, A.V., Cakir, B., Prete, M., Harun, N. and Darrigrand, J.F., 2024. Multiscale spatial mapping of cell populations across anatomical sites in healthy human skin and basal cell carcinoma. Proceedings of the National Academy of Sciences, 121(2), p.e2313326120.

## Loading in multiple donor runs
```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369529/filtered_feature_bc_matrix")
KCL9369529 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369529@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '77'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'ear'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_ear1'

##Adding metadata back into seurat object
KCL9369529 <- AddMetaData(KCL9369529, Meta)
```

```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369530/filtered_feature_bc_matrix")
KCL9369530 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369530@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '56'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Nose'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_nose1'

##Adding metadata back into seurat object
KCL9369530 <- AddMetaData(KCL9369530, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369531/filtered_feature_bc_matrix")
KCL9369531 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369531@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '85'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Cheek'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_cheek1'

##Adding metadata back into seurat object
KCL9369531 <- AddMetaData(KCL9369531, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369532/filtered_feature_bc_matrix")
KCL9369532 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369532@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '85'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Forehead'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_forehead1'

##Adding metadata back into seurat object
KCL9369532 <- AddMetaData(KCL9369532, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369533/filtered_feature_bc_matrix")
KCL9369533 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369533@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '70'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'

Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Forehead'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_forehead2'

##Adding metadata back into seurat object
KCL9369533 <- AddMetaData(KCL9369533, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369534/filtered_feature_bc_matrix")
KCL9369534 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369534@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '62'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Cheek'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_cheek2'

##Adding metadata back into seurat object
KCL9369534 <- AddMetaData(KCL9369534, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369625/filtered_feature_bc_matrix")
KCL9369625 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369625@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '86'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Ear'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_ear2'

##Adding metadata back into seurat object
KCL9369625 <- AddMetaData(KCL9369625, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369628/filtered_feature_bc_matrix")
KCL9369628 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369628@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '59'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Temple'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_temple1'

##Adding metadata back into seurat object
KCL9369628 <- AddMetaData(KCL9369628, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369629/filtered_feature_bc_matrix")
KCL9369629 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369629@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '80'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Forehead'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_forehead3'

##Adding metadata back into seurat object
KCL9369629 <- AddMetaData(KCL9369629, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369631/filtered_feature_bc_matrix")
KCL9369631 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369631@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '78'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Temple'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_temple2'

##Adding metadata back into seurat object
KCL9369631 <- AddMetaData(KCL9369631, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369632/filtered_feature_bc_matrix")
KCL9369632 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369632@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '81'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Female'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Cheek'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_cheek3b'

##Adding metadata back into seurat object
KCL9369632 <- AddMetaData(KCL9369632, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL10525738/filtered_feature_bc_matrix")
KCL10525738 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL10525738@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '90'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Forehead'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_forehead4'

##Adding metadata back into seurat object
KCL10525738 <- AddMetaData(KCL10525738, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL10525739/filtered_feature_bc_matrix")
KCL10525739 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL10525739@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '90'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Cheek'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_cheek4'

##Adding metadata back into seurat object
KCL10525739 <- AddMetaData(KCL10525739, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL10525741/filtered_feature_bc_matrix")
KCL10525741 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL10525741@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '73'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Forehead'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_forehead5'

##Adding metadata back into seurat object
KCL10525741 <- AddMetaData(KCL10525741, Meta)
```

##Preprocessing steps

### Cell-level filtering
See: [https://hbctraining.github.io/scRNA-seq_online/lessons/04_SC_quality_control.html](https://hbctraining.github.io/scRNA-seq_online/lessons/04_SC_quality_control.html)

Now that we have visualized the various metrics, we can decide on the thresholds to apply which will result in the removal of low quality cells. 

To remove potentially apoptotic or 'low quality' cells we discarded cells with less than 400 expressed genes, and/or more than 12% mitochondrial reads. We also filtered out cells with a UMI count less than 1000. 

nCount_RNA > 1000
nFeature_RNA > 400
percent.mt <= 12

To filter, we will go back to our Seurat object and use the subset() function:

```{r}

list <- list(KCL10525738, KCL10525739, KCL10525741, KCL9369529, KCL9369530, KCL9369531, KCL9369532, KCL9369533, KCL9369534, KCL9369625, KCL9369628, KCL9369629, KCL9369631, KCL9369632)

library(Seurat)
library(ggplot2)

# Define a function for preprocessing
preprocess_seurat_object <- function(seurat_object, output_path = "../Output/Ganier") {
  
  # Extract age
  age <- seurat_object@meta.data$age
  
  # Adding mitochondrial percentage and number of genes per UMI
  seurat_object$log10GenesPerUMI <- log10(as.numeric(seurat_object$nFeature_RNA) / as.numeric(seurat_object$nCount_RNA))
  seurat_object[["percent.mt"]] <- PercentageFeatureSet(seurat_object, pattern = "^MT-")
  
  # Cell-level filtering
  seurat_object <- subset(seurat_object, subset = nCount_RNA > 1000 & nFeature_RNA > 400 & percent.mt <= 12)

  # Visualize the number of cells per sample
  CellNumbers <- ggplot(seurat_object@meta.data, aes(x = age, fill = age)) +
    geom_bar() +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
    theme(plot.title = element_text(hjust = 0.5, face = "bold")) +
    ggtitle("NCells")

  ggsave(file.path(output_path, paste("NumbersPerSample_age_", age, ".tiff", sep = "")), plot = CellNumbers, device = "tiff")

  # Visualize QC metrics
  violin <- VlnPlot(seurat_object, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"))
  ggsave(file.path(output_path, paste("HDF_QC_violin_age_", age, ".tiff", sep = "")), plot = violin, device = "tiff")

  # Visualize the number of UMIs per cell (nCount_RNA)
  UMIspercell <- ggplot(seurat_object@meta.data, aes(color = age, x = nCount_RNA, fill = age)) +
    geom_density(alpha = 0.2) +
    scale_x_log10() +
    theme_classic() +
    ylab("Cell density") +
    geom_vline(xintercept = 1000) 

  ggsave(file.path(output_path, paste("UMI-per-cell_age_", age, ".tiff", sep = "")), plot = UMIspercell, device = "tiff")

  # Visualize number of genes per cell (nFeature_RNA)
  hist_feature <- ggplot(seurat_object@meta.data, aes(color = age, x = nFeature_RNA, fill = age)) +
    geom_density(alpha = 0.2) +
    theme_classic() +
    scale_x_log10() +
    geom_vline(xintercept = c(400)) 

  ggsave(file.path(output_path, paste("hist_featureRNA_age_", age, ".tiff", sep = "")), plot = hist_feature, device = "tiff")

  boxplot_feature <- ggplot(seurat_object@meta.data, aes(x = age, y = log10(nFeature_RNA), fill = age)) +
    geom_boxplot() +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
    theme(plot.title = element_text(hjust = 0.5, face = "bold")) +
    ggtitle("NCells vs NGenes")

  ggsave(file.path(output_path, paste("boxplot_featureRNA_age_", age, ".tiff", sep = "")), plot = boxplot_feature, device = "tiff")

  # Combine number of genes, UMIs, and mitochondrial percentage
  thresholding <- ggplot(seurat_object@meta.data, aes(x = nCount_RNA, y = nFeature_RNA, color = percent.mt)) +
    geom_point() +
    scale_colour_gradient(low = "gray90", high = "black") +
    stat_smooth(method = lm) +
    scale_x_log10() +
    scale_y_log10() +
    theme_classic() +
    geom_vline(xintercept = 1000) +
    geom_hline(yintercept = 400) +
    facet_wrap(~age) 

  ggsave(file.path(output_path, paste("thresholding_feature_count_mito_age_", age, ".tiff", sep = "")), plot = thresholding, device = "tiff")

  # Visualize mitochondrial gene expression per cell
  mitochondrial_genes <- ggplot(seurat_object@meta.data, aes(color = age, x = percent.mt, fill = age)) +
    geom_density(alpha = 0.2) +
    scale_x_log10() +
    theme_classic() +
    geom_vline(xintercept = 12) 

  ggsave(file.path(output_path, paste("mitochondrial_hist_age_", age, ".tiff", sep = "")), plot = mitochondrial_genes, device = "tiff")

  # Visualize the proportion of genes per UMI
  genesperUMI <- ggplot(seurat_object@meta.data, aes(x = log10GenesPerUMI, color = age, fill = age)) +
    geom_density(alpha = 0.2) +
    theme_classic() +
    geom_vline(xintercept = 0.8) 

  ggsave(file.path(output_path, paste("genesperUMI_age_", age, ".tiff", sep = "")), plot = genesperUMI, device = "tiff")

  return(seurat_object)
}

# Apply the function to each Seurat object in the list
processed_seurat_list <- lapply(list, preprocess_seurat_object)


# Removing duplicate donor samples
processed_seurat_list <- processed_seurat_list[-c(1, 6)]

```
### Normalisation
```{r}
library(Seurat)
library(ggplot2)

# Function to subset and normalize data
process_seurat_list <- function(seurat_list, output_path = "../Output/Ganier") {
  processed_seurat_list <- lapply(seurat_list, function(seurat_object) {
    
    # Subsetting Seurat object
    seurat_filt <- subset(x = seurat_object,
                          subset = percent.mt <= 12 & nFeature_RNA >= 400 & nCount_RNA >= 1000)
    
    # Visualizing number of cells before and after subsetting
    df_list <- list(Original = seurat_object@meta.data, Filtered = seurat_filt@meta.data)
    dat <- stack(lapply(df_list, `[[`, "age"))
    numbers_after_filt <- ggplot(dat, aes(x = values, fill = ind)) + geom_bar(position = "dodge")
    
    ggsave(file.path(output_path, "barplot_beforeandafterfilt.tiff"), plot = last_plot(), device = "tiff")
    
    # Normalization
    seurat_filt <- NormalizeData(seurat_filt)
    
    return(seurat_filt)
  })
  
  return(processed_seurat_list)
}

# Apply the function to all Ganier seurat objects
processed_seurat_list <- process_seurat_list(processed_seurat_list)

```
## Select highly variable features, scale, and perform PCA and UMAP dimensionality reduction
```{r}
analyze_seurat_object <- function(seurat_object, output_path = "../Output/Ganier") {
  # Extract unique ages from the metadata
  unique_ages <- unique(seurat_object@meta.data[["age"]])
  print("Unique ages in the metadata:")
  print(unique_ages)
  
  # Loop over unique ages
  for (age in unique_ages) {
    print(paste("Processing Seurat object with age:", age))

    # Rest of the function remains unchanged...
    # Selecting highly variable features
    seurat_object <- FindVariableFeatures(seurat_object, selection.method = "vst", nfeatures = 2000)
    top10 <- head(VariableFeatures(seurat_object), 10)
    plot3 <- VariableFeaturePlot(seurat_object)
    plot4 <- LabelPoints(plot = plot3, points = top10, repel = TRUE)

    ggsave(file.path(output_path, paste0("Ganier_variablefeatures_", age, ".tiff")),
           plot = last_plot(), device = "tiff")
    print("Saved variable features plot.")

    # Scaling the data and performing PCA
    all_genes <- rownames(seurat_object)
    seurat_object <- ScaleData(seurat_object, features = all_genes)
    seurat_object <- RunPCA(seurat_object, features = VariableFeatures(object = seurat_object))

    PCA_no_label <- DimPlot(seurat_object, reduction = "pca", seed = 123, combine = TRUE) + NoLegend()
    PCA_Sample_label <- DimPlot(seurat_object, reduction = "pca", seed = 123, group.by = "age")

    ggsave(file.path(output_path, paste0("PCA_no_label_", age, ".tiff")),
           plot = PCA_no_label, device = "tiff")
    ggsave(file.path(output_path, paste0("PCA_Sample_label_", age, ".tiff")),
           plot = PCA_Sample_label, device = "tiff")
    print("Saved PCA plots.")

    # Heatmaps and Elbow Plot
    DimHeatmap(seurat_object, dims = 1, cells = 500, balanced = TRUE)
    DimHeatmap(seurat_object, dims = 1:15, cells = 500, balanced = TRUE)
    ElbowPlot(seurat_object)
    print("Saved heatmaps and elbow plot.")

    # Clustering using KNN
    seurat_object <- FindNeighbors(seurat_object, dims = 1:10)
    seurat_object <- FindClusters(seurat_object, resolution = 0.4)
    head(Idents(seurat_object), 5)
    print("Performed clustering.")

    # UMAP plots
    seurat_object <- RunUMAP(seurat_object, dims = 1:10)

    UMAP_plot <- DimPlot(seurat_object, reduction = "umap", seed = 123)

    ggsave(file.path(output_path, paste0("UMAP_Ganier_", age, ".tiff")),
           plot = UMAP_plot, device = "tiff")
    print("Saved UMAP plot.")

    # Finding DEGs/cluster biomarkers
    seurat_markers <- FindAllMarkers(seurat_object, only.pos = FALSE, min.pct = 0.25, logfc.threshold = 0.25)
    write.csv(seurat_markers, file.path(output_path, paste0("Ganier_clustermarkers_posandneg_", age, ".csv")))

    seurat_markers %>%
      group_by(cluster) %>%
      top_n(n = 2, wt = avg_log2FC)
    print("Found DEGs/cluster biomarkers.")

    # Saving Seurat Object with cluster identities
    saveRDS(seurat_object, file = file.path(output_path, paste0("ganierfiltered_", age, ".rds")))
    print("Saved Seurat object.")
  }
}

lapply(processed_seurat_list, analyze_seurat_object)

```



```{r}
Ganier_56 <- readRDS("../Output/Ganier/ganierfiltered_56.rds")

Ganier_59 <- readRDS("../Output/Ganier/ganierfiltered_59.rds")

Ganier_62 <- readRDS("../Output/Ganier/ganierfiltered_62.rds")

Ganier_70 <- readRDS("../Output/Ganier/ganierfiltered_70.rds")

Ganier_73 <- readRDS("../Output/Ganier/ganierfiltered_73.rds")

Ganier_77 <- readRDS("../Output/Ganier/ganierfiltered_77.rds")

Ganier_78 <- readRDS("../Output/Ganier/ganierfiltered_78.rds")

Ganier_80 <- readRDS("../Output/Ganier/ganierfiltered_80.rds")

Ganier_81 <- readRDS("../Output/Ganier/ganierfiltered_81.rds")

Ganier_85 <- readRDS("../Output/Ganier/ganierfiltered_85.rds")

Ganier_86 <- readRDS("../Output/Ganier/ganierfiltered_86.rds")

Ganier_90 <- readRDS("../Output/Ganier/ganierfiltered_90.rds")
```

```{r}
Ganier_list <- list(
  Ganier_56,
  Ganier_59,
  Ganier_62,
  Ganier_70,
  Ganier_73,
  Ganier_77,
  Ganier_78,
  Ganier_80,
  Ganier_81,
  Ganier_85,
  Ganier_86,
  Ganier_90
)
```

#Clustering to identify fibroblasts
```{r}
analyze_seurat_objects <- function(Ganier_list, output_path = "../Output/Ganier") {
  for (Gan_object in Ganier_list) {
    # Extract age
    age <- unique(Gan_object@meta.data$age)
    print(paste("Processing Seurat object with age:", age))

    # Identify variable features and plot
    Gan_object <- FindVariableFeatures(Gan_object, selection.method = "vst", nfeatures = 2000)
    top10 <- head(VariableFeatures(Gan_object), 10)
    plot3 <- VariableFeaturePlot(Gan_object)
    plot4 <- LabelPoints(plot = plot3, points = top10, repel = TRUE)
    
    # Save Variable Features Plot
    ggsave(
      file.path(output_path, paste0("variable_features_", "Age_", age, ".tiff")),
      plot = last_plot(),
      device = "tiff",
      width = 15,  # adjust as needed
      height = 12,  # adjust as needed
      scale = 1.5  # adjust as needed
    )
    
    # Scaling and PCA
    all.genes <- rownames(Gan_object)
    Gan_object <- ScaleData(Gan_object, features = all.genes)
    Gan_object <- RunPCA(Gan_object, features = VariableFeatures(object = Gan_object))
    
    # Plot PCA
    PCA_no_label <- DimPlot(Gan_object, reduction = "pca", seed = 123, combine = TRUE) + NoLegend()
    
    # Save PCA Plot
    ggsave(
      file.path(output_path, paste0("PCA_no_label_", "Age_", age, ".tiff")),
      plot = PCA_no_label,
      device = "tiff",
      width = 15,  # adjust as needed
      height = 12,  # adjust as needed
      scale = 1.5  # adjust as needed
    )
    
    # Heatmaps
    DimHeatmap(Gan_object, dims = 1, cells = 500, balanced = TRUE)
    DimHeatmap(Gan_object, dims = 1:15, cells = 500, balanced = TRUE)
    ElbowPlot(Gan_object)
    
    # Clustering using KNN
    Gan_object <- FindNeighbors(Gan_object, dims = 1:10)
    Gan_object <- FindClusters(Gan_object, resolution = 0.4)
    head(Idents(Gan_object), 5)
    
    # Plot UMAP
    Gan_object <- RunUMAP(Gan_object, dims = 1:10)
    # Create and Save UMAP Plot
    umap_plot <- DimPlot(Gan_object, reduction = "umap", seed = 123)
    
    # Save UMAP Plot
    ggsave(
      file.path(output_path, paste0("UMAP_Ganier_", "Age_", age, ".tiff")),
      plot = last_plot(),
      device = "tiff",
      width = 15,  # adjust as needed
      height = 12,  # adjust as needed
      scale = 1.5  # adjust as needed
    )
    
    # Dotplot
    features_to_plot <- c("SARAF", "RORA", "CCL5", "NKG7", "APOD", "APOE", "CD69", "TNF", "CTLA4", "TIGIT", "C1QA", "CCL3", "HLA-DRA", "LYZ", "SFRP2", "WISP2", "TAGLN", "MYL9", "POSTN", "COL3A1", "RGS5", "STEAP4", "TM4SF1", "IFI27", "PTPRC", "SRSF7", "XCL1", "XCL2", "LY6D", "KRT1", "CD37", "BANK1", "KRT17", "PTCH1", "IL1B", "CXCL8", "TPSB2", "KIT", "PMEL", "TRYP1", "ACAN", "COL2A1", "CST3", "CPVL", "CCL21", "MMRN1", "JCHAIN", "IGHA1", "PTGDS", "ASPN", "CCL17", "CCL22", "IL8", "FCER1A", "CDH19", "SCN7A", "ACTA2", "DES", "MYLPF", "MYL1")

    DotPlot(
      Gan_object,
      features = features_to_plot,
      assay = NULL,
      cols = c("lightgrey", "blue"),
      col.min = -2.5,
      col.max = 2.5,
      dot.min = 0,
      dot.scale = 6,
      idents = NULL,
      group.by = NULL,
      split.by = NULL,
      cluster.idents = FALSE,
      scale = TRUE,
      scale.by = "radius",
      scale.min = NA,
      scale.max = NA
    ) +
    # Adjust x-axis labels in DotPlot
    theme(axis.text.x = element_text(size = 8, angle = 90))
    
    # Save DotPlot
    ggsave(
      file.path(output_path, paste0("DotPlot_Ganier_", "Age_", age, ".tiff")),
      plot = last_plot(),
      device = "tiff",
      width = 15,  # adjust as needed
      height = 12,  # adjust as needed
      scale = 1.5  # adjust as needed
    )
  }
}

lapply(Ganier_list, analyze_seurat_objects)


```


## Subset to get just fibs
```{r}
# Assuming 'Ganier_list' is your list of Seurat objects

# Install and load the Seurat package if not already installed
# install.packages("Seurat")
# library(Seurat)

# Assuming the Seurat objects are already created and stored in 'Ganier_list'

# Load required Seurat object
library(Seurat)
# Replace 'YourSeuratObjectName' with the actual name of your Seurat object
seurat_object <- Ganier_list[[1]]

# Subset cells based on seurat_cluster values
selected_clusters <- c(1, 2, 5)

# Filter cells based on the specified clusters
Ganier_56_filt <- subset(seurat_object, cells = which(seurat_object@meta.data$seurat_cluster %in% selected_clusters))

# Print the summary to check the changes
summary(Ganier_56_filt)

# Now 'filtered_seurat_object' contains only the cells with seurat_cluster values of 1, 2, or 5



# Replace 'YourSeuratObjectName' with the actual name of your Seurat object
seurat_object <- Ganier_list[[2]]

# Subset cells based on seurat_cluster values
selected_clusters <- c(1, 4)

# Filter cells based on the specified clusters
Ganier_59_filt <- subset(seurat_object, cells = which(seurat_object@meta.data$seurat_cluster %in% selected_clusters))

# Print the summary to check the changes
summary(Ganier_59_filt)

# Now 'filtered_seurat_object' contains only the cells with seurat_cluster values of 1, 2, or 5




# Replace 'YourSeuratObjectName' with the actual name of your Seurat object
seurat_object <- Ganier_list[[3]]

# Subset cells based on seurat_cluster values
selected_clusters <- c(2, 4)

# Filter cells based on the specified clusters
Ganier_62_filt <- subset(seurat_object, cells = which(seurat_object@meta.data$seurat_cluster %in% selected_clusters))

# Print the summary to check the changes
summary(Ganier_62_filt)

# Now 'filtered_seurat_object' contains only the cells with seurat_cluster values of 1, 2, or 5



# Load required Seurat object
library(Seurat)
# Replace 'YourSeuratObjectName' with the actual name of your Seurat object
seurat_object <- Ganier_list[[4]]

# Subset cells based on seurat_cluster values
selected_clusters <- c(1, 4, 8)

# Filter cells based on the specified clusters
Ganier_70_filt <- subset(seurat_object, cells = which(seurat_object@meta.data$seurat_cluster %in% selected_clusters))

# Print the summary to check the changes
summary(Ganier_70_filt)

# Now 'filtered_seurat_object' contains only the cells with seurat_cluster values of 1, 2, or 5



# Load required Seurat object
library(Seurat)
# Replace 'YourSeuratObjectName' with the actual name of your Seurat object
seurat_object <- Ganier_list[[5]]

# Subset cells based on seurat_cluster values
selected_clusters <- c(4)

# Filter cells based on the specified clusters
Ganier_73_filt <- subset(seurat_object, cells = which(seurat_object@meta.data$seurat_cluster %in% selected_clusters))

# Print the summary to check the changes
summary(Ganier_73_filt)

# Now 'filtered_seurat_object' contains only the cells with seurat_cluster values of 1, 2, or 5



# Load required Seurat object
library(Seurat)
# Replace 'YourSeuratObjectName' with the actual name of your Seurat object
seurat_object <- Ganier_list[[6]]

# Subset cells based on seurat_cluster values
selected_clusters <- c(1, 6)

# Filter cells based on the specified clusters
Ganier_77_filt <- subset(seurat_object, cells = which(seurat_object@meta.data$seurat_cluster %in% selected_clusters))

# Print the summary to check the changes
summary(Ganier_77_filt)

# Now 'filtered_seurat_object' contains only the cells with seurat_cluster values of 1, 2, or 5



# Load required Seurat object
library(Seurat)
# Replace 'YourSeuratObjectName' with the actual name of your Seurat object
seurat_object <- Ganier_list[[7]]

# Subset cells based on seurat_cluster values
selected_clusters <- c(0, 1, 5)

# Filter cells based on the specified clusters
Ganier_78_filt <- subset(seurat_object, cells = which(seurat_object@meta.data$seurat_cluster %in% selected_clusters))

# Print the summary to check the changes
summary(Ganier_78_filt)

# Now 'filtered_seurat_object' contains only the cells with seurat_cluster values of 1, 2, or 5



# Load required Seurat object
library(Seurat)
# Replace 'YourSeuratObjectName' with the actual name of your Seurat object
seurat_object <- Ganier_list[[8]]

# Subset cells based on seurat_cluster values
selected_clusters <- c(4)

# Filter cells based on the specified clusters
Ganier_80_filt <- subset(seurat_object, cells = which(seurat_object@meta.data$seurat_cluster %in% selected_clusters))

# Print the summary to check the changes
summary(Ganier_80_filt)

# Now 'filtered_seurat_object' contains only the cells with seurat_cluster values of 1, 2, or 5



# Load required Seurat object
library(Seurat)
# Replace 'YourSeuratObjectName' with the actual name of your Seurat object
seurat_object <- Ganier_list[[9]]

# Subset cells based on seurat_cluster values
selected_clusters <- c(1, 6, 7)

# Filter cells based on the specified clusters
Ganier_81_filt <- subset(seurat_object, cells = which(seurat_object@meta.data$seurat_cluster %in% selected_clusters))

# Print the summary to check the changes
summary(Ganier_81_filt)

# Now 'filtered_seurat_object' contains only the cells with seurat_cluster values of 1, 2, or 5




# Load required Seurat object
library(Seurat)
# Replace 'YourSeuratObjectName' with the actual name of your Seurat object
seurat_object <- Ganier_list[[10]]

# Subset cells based on seurat_cluster values
selected_clusters <- c(2, 4)

# Filter cells based on the specified clusters
Ganier_85_filt <- subset(seurat_object, cells = which(seurat_object@meta.data$seurat_cluster %in% selected_clusters))

# Print the summary to check the changes
summary(Ganier_85_filt)

# Now 'filtered_seurat_object' contains only the cells with seurat_cluster values of 1, 2, or 5



# Load required Seurat object
library(Seurat)
# Replace 'YourSeuratObjectName' with the actual name of your Seurat object
seurat_object <- Ganier_list[[11]]

# Subset cells based on seurat_cluster values
selected_clusters <- c(1, 4, 7)

# Filter cells based on the specified clusters
Ganier_86_filt <- subset(seurat_object, cells = which(seurat_object@meta.data$seurat_cluster %in% selected_clusters))

# Print the summary to check the changes
summary(Ganier_86_filt)

# Now 'filtered_seurat_object' contains only the cells with seurat_cluster values of 1, 2, or 5



# Load required Seurat object
library(Seurat)
# Replace 'YourSeuratObjectName' with the actual name of your Seurat object
seurat_object <- Ganier_list[[12]]

# Subset cells based on seurat_cluster values
selected_clusters <- c(2, 6)

# Filter cells based on the specified clusters
Ganier_90_filt <- subset(seurat_object, cells = which(seurat_object@meta.data$seurat_cluster %in% selected_clusters))

# Print the summary to check the changes
summary(Ganier_90_filt)

# Now 'filtered_seurat_object' contains only the cells with seurat_cluster values of 1, 2, or 5

```

## Saving seurat objects
```{r}
filt_list <- list(Ganier_56_filt, Ganier_59_filt, Ganier_62_filt, Ganier_70_filt, Ganier_73_filt, Ganier_77_filt, Ganier_78_filt, Ganier_80_filt, Ganier_81_filt, Ganier_85_filt, Ganier_86_filt, Ganier_90_filt)

save_filtered_seurat_objects <- function(seurat_list, output_dir = "../Output/FilteredGanier/") {
  # Create the output directory if it doesn't exist
  if (!file.exists(output_dir)) {
    dir.create(output_dir, recursive = TRUE)
  }

  # Loop through each Seurat object in the list
  for (i in seq_along(seurat_list)) {
    seurat_object <- seurat_list[[i]]
    
    # Extract age value from @meta.data
    age_value <- unique(seurat_object@meta.data$age)

    # Replace spaces with underscores and remove special characters
    sanitized_age <- gsub("[[:space:]]", "_", age_value)
    sanitized_age <- gsub("[^[:alnum:]_]", "", sanitized_age)

    # Generate the output filename
    output_filename <- paste0(output_dir, "filteredseurat_Ganier_", sanitized_age, ".rds")

    # Save the filtered Seurat object
    saveRDS(seurat_object, file = output_filename)
    
    cat("Saved:", output_filename, "\n")
  }
}

# Example usage assuming 'filt_list' is your list of filtered Seurat objects
save_filtered_seurat_objects(filt_list)

```


```{r}
Ganier_56 <- readRDS("../Output/FilteredGanier/filteredseurat_Ganier_56.rds")

Ganier_59 <- readRDS("../Output/FilteredGanier/filteredseurat_Ganier_59.rds")

Ganier_62 <- readRDS("../Output/FilteredGanier/filteredseurat_Ganier_62.rds")

Ganier_70 <- readRDS("../Output/FilteredGanier/filteredseurat_Ganier_70.rds")

Ganier_73 <- readRDS("../Output/FilteredGanier/filteredseurat_Ganier_73.rds")

Ganier_77 <- readRDS("../Output/FilteredGanier/filteredseurat_Ganier_77.rds")

Ganier_78 <- readRDS("../Output/FilteredGanier/filteredseurat_Ganier_78.rds")

Ganier_80 <- readRDS("../Output/FilteredGanier/filteredseurat_Ganier_80.rds")

Ganier_81 <- readRDS("../Output/FilteredGanier/filteredseurat_Ganier_81.rds")

Ganier_85 <- readRDS("../Output/FilteredGanier/filteredseurat_Ganier_85.rds")

Ganier_86 <- readRDS("../Output/FilteredGanier/filteredseurat_Ganier_86.rds")

Ganier_90 <- readRDS("../Output/FilteredGanier/filteredseurat_Ganier_90.rds")
```

```{r}
Ganier_list <- list(
  Ganier_56,
  Ganier_59,
  Ganier_62,
  Ganier_70,
  Ganier_73,
  Ganier_77,
  Ganier_78,
  Ganier_80,
  Ganier_81,
  Ganier_85,
  Ganier_86,
  Ganier_90
)
```


```{r}
library(Seurat)

# Assuming 'Ganier_list' is your list of Seurat objects

# Define a function to run the code and save barplot for each Seurat object
run_and_save_barplot <- function(seurat_object) {
  # NormalizeData and scPredict
  query <- NormalizeData(seurat_object)
  query <- scPredict(query, hdf_mda3D)

  # Extract metadata for further analysis
  query2 <- query@meta.data
  table <- table(query2[, c("age", "scpred_prediction")])
  mat <- as.data.frame.matrix(table)
  
  # Calculate percentages and create barplot
  percentages <- (mat / rowSums(mat)) * 100
  percentages <- na.omit(percentages)
  percentages <- percentages[order(row.names(percentages)), ]
  v <- rownames(percentages)

  # Save the percentages data for each Seurat object
  percentages_file <- file.path("../Output/Ganier", paste0("Percentages_DS3D_by_age_", seurat_object@meta.data$age, ".csv"))
  write.csv(percentages, file = percentages_file)

  # Create barplot
  barplot_data <- percentages$DS3D
  barplot_names <- v

  barplot_file <- file.path("../Output/Ganier", paste0("scPred3D_barplot_DS3Dbyage_", seurat_object@meta.data$age, ".png"))
  png(filename = barplot_file, width = 800, height = 600)
  barplot(barplot_data, names = barplot_names, cex.names = 0.7, col = "darkblue",
          main = "Percentage of predicted deeply senescent cells", xlab = "Age",
          ylab = "Percentage", ylim = c(0, 100))
  dev.off()
}

# Apply the function to your list of Seurat objects
lapply(Ganier_list, run_and_save_barplot)


```


#Combining Tabib, Sole, and Ganier predictions for scatterplot - 3D

```{r}
Tabib <- read.csv("../Output/Tabib/EPDS3D_tabibfibs.csv")
Sole <- read.csv("../Output/SoleBoldo/EPDS3D_solefibs.csv")
Ganier56 <- read.csv("../Output/Ganier/Percentages_DS3D_by_age_56.csv")
Ganier59 <- read.csv("../Output/Ganier/Percentages_DS3D_by_age_59.csv")
Ganier62 <- read.csv("../Output/Ganier/Percentages_DS3D_by_age_62.csv")
Ganier70 <- read.csv("../Output/Ganier/Percentages_DS3D_by_age_70.csv")
Ganier73 <- read.csv("../Output/Ganier/Percentages_DS3D_by_age_73.csv")
Ganier77 <- read.csv("../Output/Ganier//Percentages_DS3D_by_age_77.csv")
Ganier78 <- read.csv("../Output/Ganier/Percentages_DS3D_by_age_78.csv")
Ganier81 <- read.csv("../Output/Ganier/Percentages_DS3D_by_age_81.csv")
Ganier90 <- read.csv("../Output/Ganier/Percentages_DS3D_by_age_90.csv")

# Set the first column as row names
rownames(Ganier56) <- Ganier56[, 1]

# Remove the first column (it is now redundant)
Ganier56 <- Ganier56[, -1]

# Set the first column as row names
rownames(Ganier59) <- Ganier59[, 1]

# Remove the first column (it is now redundant)
Ganier59 <- Ganier59[, -1]

# Set the first column as row names
rownames(Ganier62) <- Ganier62[, 1]

# Remove the first column (it is now redundant)
Ganier62 <- Ganier62[, -1]

# Set the first column as row names
rownames(Ganier70) <- Ganier70[, 1]

# Remove the first column (it is now redundant)
Ganier70 <- Ganier70[, -1]

# Set the first column as row names
rownames(Ganier73) <- Ganier73[, 1]

# Remove the first column (it is now redundant)
Ganier73 <- Ganier73[, -1]

# Set the first column as row names
rownames(Ganier77) <- Ganier77[, 1]

# Remove the first column (it is now redundant)
Ganier77 <- Ganier77[, -1]

# Set the first column as row names
rownames(Ganier78) <- Ganier78[, 1]

# Remove the first column (it is now redundant)
Ganier78 <- Ganier78[, -1]

# Set the first column as row names
rownames(Ganier81) <- Ganier81[, 1]

# Remove the first column (it is now redundant)
Ganier81 <- Ganier81[, -1]


rownames(Ganier90) <- Ganier90[, 1]

# Remove the first column (it is now redundant)
Ganier90 <- Ganier90[, -1]

# Set the first column as row names
rownames(Tabib) <- Tabib[, 1]

# Remove the first column (it is now redundant)
Tabib <- Tabib[, -1]

# Set the first column as row names
rownames(Sole) <- Sole[, 1]

# Remove the first column (it is now redundant)
Sole <- Sole[, -1]

```


```{r}
# Create a list of data frames
df_list <- list(Tabib, Sole, Ganier56, Ganier59, Ganier62, Ganier70, Ganier73, Ganier77, Ganier78, Ganier81, Ganier90)

# Extract row names and create a new column "age" in each data frame
for (i in seq_along(df_list)) {
  df_list[[i]]$age <- rownames(df_list[[i]])
}

# Merge data frames based on common columns
merged_df <- Reduce(function(x, y) merge(x, y, by = intersect(names(x), names(y)), all = TRUE), df_list)

# Print the merged data frame
print(merged_df)

file_path <- "../Output/allpercentagepredictions_fibs3D.csv"
write.csv(merged_df, file = file_path, row.names = TRUE)
```

```{r}
merged_df <- merged_df[, -c(2, 4)]

# Assuming merged_df is your merged data frame
library(ggplot2)

# Convert merged_df to long format
merged_long <- tidyr::pivot_longer(merged_df, cols = -age, names_to = "variable", values_to = "value")

# Convert 'age' to numeric
merged_long$age <- as.numeric(merged_long$age)

png(filename = "../Output/percentageDSpredicted_all_fibs_3D.png", width = 800, height = 600)  # Set appropriate width and height

# Create a scatterplot with a trendline
p <- ggplot(merged_long, aes(x = age, y = value)) +
  geom_point(size = 3, color = "blue") +  # Make points larger and blue
  geom_smooth(method = "lm", se = FALSE, color = "black", size = 0.7) +  # Make trendline slightly thinner
  labs(title = NULL, x = "Age", y = "Percentage DS predicted cells") +
  theme_minimal() +
  theme(panel.grid = element_blank())  # Remove gridlines

# Extract R-squared and p-values
fit <- lm(value ~ age, data = merged_long)
rsq <- sprintf("R^2 = %.3f", summary(fit)$r.squared)
pval <- sprintf("p = %.3f", summary(fit)$coefficients[2,4])

# Add R-squared and p-values as text
p + 
  geom_text(x = max(merged_long$age), y = max(merged_long$value), label = paste(rsq, pval, sep = ", "), hjust = 1, vjust = 10)

dev.off()

```

# Testing the SenPred 2D EP/DS model on Ganier, and merging the predictions into a scatterplot with Tabib and Sole-Boldo data
```{r}
library(Seurat)

# Assuming 'Ganier_list' is your list of Seurat objects

# Define a function to run the code and save barplot for each Seurat object
run_and_save_barplot <- function(seurat_object) {
  # NormalizeData and scPredict
  query <- NormalizeData(seurat_object)
  query <- scPredict(query, hdf_mda2D)

  # Extract metadata for further analysis
  query2 <- query@meta.data
  table <- table(query2[, c("age", "scpred_prediction")])
  mat <- as.data.frame.matrix(table)
  
  # Calculate percentages and create barplot
  percentages <- (mat / rowSums(mat)) * 100
  percentages <- na.omit(percentages)
  percentages <- percentages[order(row.names(percentages)), ]
  v <- rownames(percentages)

  # Save the percentages data for each Seurat object
  percentages_file <- file.path("../Output/Ganier", paste0("Percentages_DS2D_by_age_", seurat_object@meta.data$age, ".csv"))
  write.csv(percentages, file = percentages_file)

  # Create barplot
  barplot_data <- percentages$DS3D
  barplot_names <- v

  barplot_file <- file.path("../Output/Ganier", paste0("scPred3D_barplot_DS2Dbyage_", seurat_object@meta.data$age, ".png"))
  png(filename = barplot_file, width = 800, height = 600)
  barplot(barplot_data, names = barplot_names, cex.names = 0.7, col = "darkblue",
          main = "Percentage of predicted deeply senescent cells", xlab = "Age",
          ylab = "Percentage", ylim = c(0, 100))
  dev.off()
}

# Apply the function to your list of Seurat objects
lapply(Ganier_list, run_and_save_barplot)


```


#Combining Tabib, Sole, and Ganier predictions for scatterplot - 3D

```{r}
Tabib <- read.csv("../Output/Tabib/EPDS2D_tabibfibs.csv")
Sole <- read.csv("../Output/SoleBoldo/EPDS2D_solefibs.csv")
Ganier56 <- read.csv("../Output/Ganier/Percentages_DS2D_by_age_56.csv")
Ganier59 <- read.csv("../Output/Ganier/Percentages_DS2D_by_age_59.csv")
Ganier62 <- read.csv("../Output/Ganier/Percentages_DS2D_by_age_62.csv")
Ganier70 <- read.csv("../Output/Ganier/Percentages_DS2D_by_age_70.csv")
Ganier73 <- read.csv("../Output/Ganier/Percentages_DS2D_by_age_73.csv")
Ganier77 <- read.csv("../Output/Ganier//Percentages_DS2D_by_age_77.csv")
Ganier78 <- read.csv("../Output/Ganier/Percentages_DS2D_by_age_78.csv")
Ganier81 <- read.csv("../Output/Ganier/Percentages_DS2D_by_age_81.csv")
Ganier90 <- read.csv("../Output/Ganier/Percentages_DS2D_by_age_90.csv")

# Set the first column as row names
rownames(Ganier56) <- Ganier56[, 1]

# Remove the first column (it is now redundant)
Ganier56 <- Ganier56[, -1]

# Set the first column as row names
rownames(Ganier59) <- Ganier59[, 1]

# Remove the first column (it is now redundant)
Ganier59 <- Ganier59[, -1]

# Set the first column as row names
rownames(Ganier62) <- Ganier62[, 1]

# Remove the first column (it is now redundant)
Ganier62 <- Ganier62[, -1]

# Set the first column as row names
rownames(Ganier70) <- Ganier70[, 1]

# Remove the first column (it is now redundant)
Ganier70 <- Ganier70[, -1]

# Set the first column as row names
rownames(Ganier73) <- Ganier73[, 1]

# Remove the first column (it is now redundant)
Ganier73 <- Ganier73[, -1]

# Set the first column as row names
rownames(Ganier77) <- Ganier77[, 1]

# Remove the first column (it is now redundant)
Ganier77 <- Ganier77[, -1]

# Set the first column as row names
rownames(Ganier78) <- Ganier78[, 1]

# Remove the first column (it is now redundant)
Ganier78 <- Ganier78[, -1]

# Set the first column as row names
rownames(Ganier81) <- Ganier81[, 1]

# Remove the first column (it is now redundant)
Ganier81 <- Ganier81[, -1]


rownames(Ganier90) <- Ganier90[, 1]

# Remove the first column (it is now redundant)
Ganier90 <- Ganier90[, -1]

# Set the first column as row names
rownames(Tabib) <- Tabib[, 1]

# Remove the first column (it is now redundant)
Tabib <- Tabib[, -1]

# Set the first column as row names
rownames(Sole) <- Sole[, 1]

# Remove the first column (it is now redundant)
Sole <- Sole[, -1]

```


```{r}
# Create a list of data frames
df_list <- list(Tabib, Sole, Ganier56, Ganier59, Ganier62, Ganier70, Ganier73, Ganier77, Ganier78, Ganier81, Ganier90)

# Extract row names and create a new column "age" in each data frame
for (i in seq_along(df_list)) {
  df_list[[i]]$age <- rownames(df_list[[i]])
}

# Merge data frames based on common columns
merged_df <- Reduce(function(x, y) merge(x, y, by = intersect(names(x), names(y)), all = TRUE), df_list)

# Print the merged data frame
print(merged_df)

file_path <- "../Output/allpercentagepredictions_fibs2D.csv"
write.csv(merged_df, file = file_path, row.names = TRUE)
```

```{r}
merged_df <- merged_df[, -c(2, 4)]

# Assuming merged_df is your merged data frame
library(ggplot2)

# Convert merged_df to long format
merged_long <- tidyr::pivot_longer(merged_df, cols = -age, names_to = "variable", values_to = "value")

# Convert 'age' to numeric
merged_long$age <- as.numeric(merged_long$age)

png(filename = "../Output/percentageDSpredicted_all_fibs_2D.png", width = 800, height = 600)  # Set appropriate width and height

# Create a scatterplot with a trendline
p <- ggplot(merged_long, aes(x = age, y = value)) +
  geom_point(size = 3, color = "blue") +  # Make points larger and blue
  geom_smooth(method = "lm", se = FALSE, color = "black", size = 0.7) +  # Make trendline slightly thinner
  labs(title = NULL, x = "Age", y = "Percentage DS predicted cells") +
  theme_minimal() +
  theme(panel.grid = element_blank())  # Remove gridlines

# Extract R-squared and p-values
fit <- lm(value ~ age, data = merged_long)
rsq <- sprintf("R^2 = %.3f", summary(fit)$r.squared)
pval <- sprintf("p = %.3f", summary(fit)$coefficients[2,4])

# Add R-squared and p-values as text
p + 
  geom_text(x = max(merged_long$age), y = max(merged_long$value), label = paste(rsq, pval, sep = ", "), hjust = 1, vjust = 10)

dev.off()

```

### Session information
```{r session_info_99}
sessionInfo()
```

