---
title: "16_Ganier"
author: "BH"
date: "2024-03-21"
output:
 pdf_document:
    toc: yes
---
```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
```

```{r load_packages}
library("dplyr")
library("Seurat")
library("patchwork")
library("readxl")
library("tidyseurat")
library("ggplot2")
library("tidyverse")
library("RCurl")
library("cowplot")
library("sctransform")
library("scPred")
library("Seurat")
library("magrittr")
library("caret")
library("dplyr")
library("janitor")
library("factoextra")
library("reshape2")

options(future.globals.maxSize = 12000 * 1024^2)

```

# Adding an additional whole skin dataset from Ganier et al. This data was downloaded from ArrayExpress under “E- MTAB- 13085”, from the following paper:
Ganier, C., Mazin, P., Herrera-Oropeza, G., Du-Harpur, X., Blakeley, M., Gabriel, J., Predeus, A.V., Cakir, B., Prete, M., Harun, N. and Darrigrand, J.F., 2024. Multiscale spatial mapping of cell populations across anatomical sites in healthy human skin and basal cell carcinoma. Proceedings of the National Academy of Sciences, 121(2), p.e2313326120.

## Loading in multiple donor runs
```{r}
custom_theme <- readRDS(file = "custom_theme.rds")
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369529/filtered_feature_bc_matrix")
KCL9369529 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369529@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '77'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'ear'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_ear1'

##Adding metadata back into seurat object
KCL9369529 <- AddMetaData(KCL9369529, Meta)
```

```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369530/filtered_feature_bc_matrix")
KCL9369530 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369530@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '56'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Nose'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_nose1'

##Adding metadata back into seurat object
KCL9369530 <- AddMetaData(KCL9369530, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369531/filtered_feature_bc_matrix")
KCL9369531 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369531@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '85'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Cheek'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_cheek1'

##Adding metadata back into seurat object
KCL9369531 <- AddMetaData(KCL9369531, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369532/filtered_feature_bc_matrix")
KCL9369532 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369532@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '85'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Forehead'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_forehead1'

##Adding metadata back into seurat object
KCL9369532 <- AddMetaData(KCL9369532, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369533/filtered_feature_bc_matrix")
KCL9369533 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369533@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '70'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'

Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Forehead'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_forehead2'

##Adding metadata back into seurat object
KCL9369533 <- AddMetaData(KCL9369533, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369534/filtered_feature_bc_matrix")
KCL9369534 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369534@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '62'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Cheek'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_cheek2'

##Adding metadata back into seurat object
KCL9369534 <- AddMetaData(KCL9369534, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369625/filtered_feature_bc_matrix")
KCL9369625 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369625@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '86'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Ear'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_ear2'

##Adding metadata back into seurat object
KCL9369625 <- AddMetaData(KCL9369625, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369628/filtered_feature_bc_matrix")
KCL9369628 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369628@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '59'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Temple'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_temple1'

##Adding metadata back into seurat object
KCL9369628 <- AddMetaData(KCL9369628, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369629/filtered_feature_bc_matrix")
KCL9369629 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369629@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '80'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Forehead'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_forehead3'

##Adding metadata back into seurat object
KCL9369629 <- AddMetaData(KCL9369629, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369631/filtered_feature_bc_matrix")
KCL9369631 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369631@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '78'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Temple'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_temple2'

##Adding metadata back into seurat object
KCL9369631 <- AddMetaData(KCL9369631, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL9369632/filtered_feature_bc_matrix")
KCL9369632 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL9369632@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '81'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Female'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Cheek'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_cheek3b'

##Adding metadata back into seurat object
KCL9369632 <- AddMetaData(KCL9369632, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL10525738/filtered_feature_bc_matrix")
KCL10525738 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL10525738@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '90'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Forehead'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_forehead4'

##Adding metadata back into seurat object
KCL10525738 <- AddMetaData(KCL10525738, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL10525739/filtered_feature_bc_matrix")
KCL10525739 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL10525739@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '90'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Cheek'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_cheek4'

##Adding metadata back into seurat object
KCL10525739 <- AddMetaData(KCL10525739, Meta)
```


```{r}
Ganier <- Read10X(data.dir= "../Inputfiles/Ganier/WS_SKN_KCL10525741/filtered_feature_bc_matrix")
KCL10525741 <- CreateSeuratObject(counts = Ganier, min.cells = 3, min.features = 200)

Meta <- KCL10525741@meta.data
Meta$organism <- 'Homo Sapiens'
Meta$age <- '73'
Meta$developmental_stage <- 'Adult'
Meta$sex <- 'Male'
Meta$organism_part <- 'Skin'
Meta$sampling_site <- 'Forehead'
Meta$disease <- 'normal'
Meta$sample_id <- 'face_forehead5'

##Adding metadata back into seurat object
KCL10525741 <- AddMetaData(KCL10525741, Meta)
```

##Preprocessing steps

### Cell-level filtering
See: [https://hbctraining.github.io/scRNA-seq_online/lessons/04_SC_quality_control.html](https://hbctraining.github.io/scRNA-seq_online/lessons/04_SC_quality_control.html)

Now that we have visualized the various metrics, we can decide on the thresholds to apply which will result in the removal of low quality cells. 

To remove potentially apoptotic or 'low quality' cells we discarded cells with less than 400 expressed genes, and/or more than 12% mitochondrial reads. We also filtered out cells with a UMI count less than 1000. 

nCount_RNA > 1000
nFeature_RNA > 400
percent.mt <= 12

To filter, we will go back to our Seurat object and use the subset() function:

```{r}

list <- list(KCL10525738, KCL10525739, KCL10525741, KCL9369529, KCL9369530, KCL9369531, KCL9369532, KCL9369533, KCL9369534, KCL9369625, KCL9369628, KCL9369629, KCL9369631, KCL9369632)

library(Seurat)
library(ggplot2)

# Define a function for preprocessing
preprocess_seurat_object <- function(seurat_object, output_path = "../Output/Ganier") {
  
  # Extract age
  age <- seurat_object@meta.data$age
  
  # Adding mitochondrial percentage and number of genes per UMI
  seurat_object$log10GenesPerUMI <- log10(as.numeric(seurat_object$nFeature_RNA) / as.numeric(seurat_object$nCount_RNA))
  seurat_object[["percent.mt"]] <- PercentageFeatureSet(seurat_object, pattern = "^MT-")
  
  # Cell-level filtering
  seurat_object <- subset(seurat_object, subset = nCount_RNA > 1000 & nFeature_RNA > 400 & percent.mt <= 12)

  # Visualize the number of cells per sample
  CellNumbers <- ggplot(seurat_object@meta.data, aes(x = age, fill = age)) +
    geom_bar() +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
    theme(plot.title = element_text(hjust = 0.5, face = "bold")) +
    ggtitle("NCells")

  ggsave(file.path(output_path, paste("NumbersPerSample_age_", age, ".tiff", sep = "")), plot = CellNumbers, device = "tiff")

  # Visualize QC metrics
  violin <- VlnPlot(seurat_object, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"))
  ggsave(file.path(output_path, paste("HDF_QC_violin_age_", age, ".tiff", sep = "")), plot = violin, device = "tiff")

  # Visualize the number of UMIs per cell (nCount_RNA)
  UMIspercell <- ggplot(seurat_object@meta.data, aes(color = age, x = nCount_RNA, fill = age)) +
    geom_density(alpha = 0.2) +
    scale_x_log10() +
    theme_classic() +
    ylab("Cell density") +
    geom_vline(xintercept = 1000) 

  ggsave(file.path(output_path, paste("UMI-per-cell_age_", age, ".tiff", sep = "")), plot = UMIspercell, device = "tiff")

  # Visualize number of genes per cell (nFeature_RNA)
  hist_feature <- ggplot(seurat_object@meta.data, aes(color = age, x = nFeature_RNA, fill = age)) +
    geom_density(alpha = 0.2) +
    theme_classic() +
    scale_x_log10() +
    geom_vline(xintercept = c(400)) 

  ggsave(file.path(output_path, paste("hist_featureRNA_age_", age, ".tiff", sep = "")), plot = hist_feature, device = "tiff")

  boxplot_feature <- ggplot(seurat_object@meta.data, aes(x = age, y = log10(nFeature_RNA), fill = age)) +
    geom_boxplot() +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
    theme(plot.title = element_text(hjust = 0.5, face = "bold")) +
    ggtitle("NCells vs NGenes")

  ggsave(file.path(output_path, paste("boxplot_featureRNA_age_", age, ".tiff", sep = "")), plot = boxplot_feature, device = "tiff")

  # Combine number of genes, UMIs, and mitochondrial percentage
  thresholding <- ggplot(seurat_object@meta.data, aes(x = nCount_RNA, y = nFeature_RNA, color = percent.mt)) +
    geom_point() +
    scale_colour_gradient(low = "gray90", high = "black") +
    stat_smooth(method = lm) +
    scale_x_log10() +
    scale_y_log10() +
    theme_classic() +
    geom_vline(xintercept = 1000) +
    geom_hline(yintercept = 400) +
    facet_wrap(~age) 

  ggsave(file.path(output_path, paste("thresholding_feature_count_mito_age_", age, ".tiff", sep = "")), plot = thresholding, device = "tiff")

  # Visualize mitochondrial gene expression per cell
  mitochondrial_genes <- ggplot(seurat_object@meta.data, aes(color = age, x = percent.mt, fill = age)) +
    geom_density(alpha = 0.2) +
    scale_x_log10() +
    theme_classic() +
    geom_vline(xintercept = 12) 

  ggsave(file.path(output_path, paste("mitochondrial_hist_age_", age, ".tiff", sep = "")), plot = mitochondrial_genes, device = "tiff")

  # Visualize the proportion of genes per UMI
  genesperUMI <- ggplot(seurat_object@meta.data, aes(x = log10GenesPerUMI, color = age, fill = age)) +
    geom_density(alpha = 0.2) +
    theme_classic() +
    geom_vline(xintercept = 0.8) 

  ggsave(file.path(output_path, paste("genesperUMI_age_", age, ".tiff", sep = "")), plot = genesperUMI, device = "tiff")

  return(seurat_object)
}

# Apply the function to each Seurat object in the list
processed_seurat_list <- lapply(list, preprocess_seurat_object)


# Removing duplicate donor samples
processed_seurat_list <- processed_seurat_list[-c(1, 6)]

```
### Normalisation
```{r}
library(Seurat)
library(ggplot2)

# Function to subset and normalize data
process_seurat_list <- function(seurat_list, output_path = "../Output/Ganier") {
  processed_seurat_list <- lapply(seurat_list, function(seurat_object) {
    
    # Subsetting Seurat object
    seurat_filt <- subset(x = seurat_object,
                          subset = percent.mt <= 12 & nFeature_RNA >= 400 & nCount_RNA >= 1000)
    
    # Visualizing number of cells before and after subsetting
    df_list <- list(Original = seurat_object@meta.data, Filtered = seurat_filt@meta.data)
    dat <- stack(lapply(df_list, `[[`, "age"))
    numbers_after_filt <- ggplot(dat, aes(x = values, fill = ind)) + geom_bar(position = "dodge")
    
    ggsave(file.path(output_path, "barplot_beforeandafterfilt.tiff"), plot = last_plot(), device = "tiff")
    
    # Normalization
    seurat_filt <- NormalizeData(seurat_filt)
    
    return(seurat_filt)
  })
  
  return(processed_seurat_list)
}

# Apply the function to all Ganier seurat objects
processed_seurat_list <- process_seurat_list(processed_seurat_list)

```
## Select highly variable features, scale, and perform PCA and UMAP dimensionality reduction
```{r}
analyze_seurat_object <- function(seurat_object, output_path = "../Output/Ganier") {
  # Extract unique ages from the metadata
  unique_ages <- unique(seurat_object@meta.data[["age"]])
  print("Unique ages in the metadata:")
  print(unique_ages)
  
  # Loop over unique ages
  for (age in unique_ages) {
    print(paste("Processing Seurat object with age:", age))

    # Rest of the function remains unchanged...
    # Selecting highly variable features
    seurat_object <- FindVariableFeatures(seurat_object, selection.method = "vst", nfeatures = 2000)
    top10 <- head(VariableFeatures(seurat_object), 10)
    plot3 <- VariableFeaturePlot(seurat_object)
    plot4 <- LabelPoints(plot = plot3, points = top10, repel = TRUE)

    ggsave(file.path(output_path, paste0("Ganier_variablefeatures_", age, ".tiff")),
           plot = last_plot(), device = "tiff")
    print("Saved variable features plot.")

    # Scaling the data and performing PCA
    all_genes <- rownames(seurat_object)
    seurat_object <- ScaleData(seurat_object, features = all_genes)
    seurat_object <- RunPCA(seurat_object, features = VariableFeatures(object = seurat_object))

    PCA_no_label <- DimPlot(seurat_object, reduction = "pca", seed = 123, combine = TRUE) + NoLegend()
    PCA_Sample_label <- DimPlot(seurat_object, reduction = "pca", seed = 123, group.by = "age")

    ggsave(file.path(output_path, paste0("PCA_no_label_", age, ".tiff")),
           plot = PCA_no_label, device = "tiff")
    ggsave(file.path(output_path, paste0("PCA_Sample_label_", age, ".tiff")),
           plot = PCA_Sample_label, device = "tiff")
    print("Saved PCA plots.")

    # Heatmaps and Elbow Plot
    DimHeatmap(seurat_object, dims = 1, cells = 500, balanced = TRUE)
    DimHeatmap(seurat_object, dims = 1:15, cells = 500, balanced = TRUE)
    ElbowPlot(seurat_object)
    print("Saved heatmaps and elbow plot.")

    # Clustering using KNN
    seurat_object <- FindNeighbors(seurat_object, dims = 1:10)
    seurat_object <- FindClusters(seurat_object, resolution = 0.4)
    head(Idents(seurat_object), 5)
    print("Performed clustering.")

    # UMAP plots
    seurat_object <- RunUMAP(seurat_object, dims = 1:10)

    UMAP_plot <- DimPlot(seurat_object, reduction = "umap", seed = 123)

    ggsave(file.path(output_path, paste0("UMAP_Ganier_", age, ".tiff")),
           plot = UMAP_plot, device = "tiff")
    print("Saved UMAP plot.")

    # Finding DEGs/cluster biomarkers
    seurat_markers <- FindAllMarkers(seurat_object, only.pos = FALSE, min.pct = 0.25, logfc.threshold = 0.25)
    write.csv(seurat_markers, file.path(output_path, paste0("Ganier_clustermarkers_posandneg_", age, ".csv")))

    seurat_markers %>%
      group_by(cluster) %>%
      top_n(n = 2, wt = avg_log2FC)
    print("Found DEGs/cluster biomarkers.")

    # Saving Seurat Object with cluster identities
    saveRDS(seurat_object, file = file.path(output_path, paste0("ganierfiltered_", age, ".rds")))
    print("Saved Seurat object.")
  }
}

lapply(processed_seurat_list, analyze_seurat_object)

```



### Session information
```{r session_info_99}
sessionInfo()
```

