---
title: "2_HDF_filtering_normalisation"
author: "Beth"
date: "24/09/2021"
output: html_document
toc: TRUE
---

## Cell-level filtering
See: [https://hbctraining.github.io/scRNA-seq_online/lessons/04_SC_quality_control.html](https://hbctraining.github.io/scRNA-seq_online/lessons/04_SC_quality_control.html)

Now that we have visualized the various metrics, we can decide on the thresholds to apply which will result in the removal of low quality cells. 

To remove potentially apoptotic or 'low quality' cells we discarded cells with less than 400 expressed genes, and/or more than 12% mitochondrial reads. We also filtered out cells with a UMI count less than 1000. 

nCount_RNA > 1000
nFeature_RNA > 400
percent.mt <= 12

To filter, we wil go back to our Seurat object and use the subset() function:


```{r}
#Note - mitochondrial percentage has been added to metadata in previous script 1_HDF_data.Rmd. Adding no. of genes per UMI.
hdfseurat$log10GenesPerUMI <- log10(hdfseurat$nFeature_RNA) / log10(hdfseurat$nCount_RNA)

#Visualise the number of cells per sample
CellNumbers <- hdfseurat@meta.data %>% 
  	           ggplot(aes(x=Sample, fill=Sample)) + 
  	           geom_bar() +
  	           theme_classic() +
  	           theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	           theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	           ggtitle("NCells")

CellNumbers

#Save plot
ggsave(
  "NumbersPerSample.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/HDF-Data-QC"
  )

```
## Visualising QC metrics - nFeature_RNA, nCount_RNA, percent.mt
```{r}
# Create plot
violin <- VlnPlot(hdfseurat, features =c("nFeature_RNA", "nCount_RNA", "percent.mt"))
violin

#Save plot
ggsave(
  "HDF_QC_violin.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/HDF-Data-QC"
  )
```

## Visualising the number of UMIs per cell (nCount_RNA)
```{r}
#Number of UMIs (transcripts) per cell
UMIspercell <- hdfseurat@meta.data %>% 
  	           ggplot(aes(color=Sample, x=nCount_RNA, fill= Sample)) + 
  	           geom_density(alpha = 0.2) + 
  	           scale_x_log10() + 
  	           theme_classic() +
  	           ylab("Cell density") +
  	           geom_vline(xintercept = 1000) 

UMIspercell

#Save plot
ggsave(
  "UMI-per-cell.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/HDF-Data-QC"
  )

```
## Visualising number of genes per cell (nFeature_RNA)
```{r}
# Visualize the distribution of genes detected per cell via histogram
hist_feature<- hdfseurat@meta.data %>% 
  	           ggplot(aes(color=Sample, x=nFeature_RNA, fill= Sample)) + 
  	           geom_density(alpha = 0.2) + 
  	           theme_classic() +
  	           scale_x_log10() + 
  	           geom_vline(xintercept = c(400)) 

hist_feature

#Save plot
ggsave(
  "hist_featureRNA.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/HDF-Data-QC"
  )

# Visualize the distribution of genes detected per cell via boxplot
boxplot_feature <- hdfseurat@meta.data %>% 
    	             ggplot(aes(x=Sample, y=log10(nFeature_RNA), fill=Sample)) + 
  	               geom_boxplot() + 
  	               theme_classic() +
  	               theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	               theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	               ggtitle("NCells vs NGenes")

boxplot_feature

#Save plot
ggsave(
  "boxplot_featureRNA.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/HDF-Data-QC"
  )
    
```

## Combining number of genes, UMIs and mitochondrial percentage, to show thresholding boundaries
```{r}
thresholding <- hdfseurat@meta.data %>% 
  	            ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percent.mt)) + 
  	            geom_point() + 
	              scale_colour_gradient(low = "gray90", high = "black") +
  	            stat_smooth(method=lm) +
  	            scale_x_log10() + 
  	            scale_y_log10() + 
  	            theme_classic() +
  	            geom_vline(xintercept = 1000) +
  	            geom_hline(yintercept = 400) +
  	            facet_wrap(~Sample) 

thresholding

ggsave(
  "thresholding_feature_count_mito.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/HDF-Data-QC"
  )
```
## Mitochondrial gene expression per cell
```{r}
# Visualize the distribution of mitochondrial gene expression detected per cell
mitochondrial_genes <- hdfseurat@meta.data %>% 
                    	 ggplot(aes(color=Sample, x=percent.mt, fill=Sample)) + 
  	                   geom_density(alpha = 0.2) + 
  	                   scale_x_log10() + 
  	                   theme_classic() +
  	                   geom_vline(xintercept = 12) 

mitochondrial_genes

ggsave(
  "mitochondrial_hist.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/HDF-Data-QC"
  )
```
## Visualising the proportion of genes per UMI
```{r}
genesperUMI <- hdfseurat@meta.data %>%
  	           ggplot(aes(x=log10GenesPerUMI, color = Sample, fill=Sample)) +
  	           geom_density(alpha = 0.2) +
  	           theme_classic() +
  	           geom_vline(xintercept = 0.8) 

genesperUMI

ggsave(
  "genesperUMI.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/HDF-Data-QC"
  )
```

#Subsetting seurat object
```{r}
hdfseurat_filt <- subset(x=hdfseurat, 
                         subset= (percent.mt <=12) &
                            (nFeature_RNA >=400) &
                            (nCount_RNA >= 1000))

#Visualising number of cells before and after subsetting
df.list <- list(Original=hdfseurat@meta.data, Filtered=hdfseurat_filt@meta.data)
dat <- stack(lapply(df.list, `[[`, "Sample"))
numbers_after_filt <- ggplot(dat, aes(x=values, fill=ind)) + geom_bar(position="dodge") 

numbers_after_filt

ggsave(
  "barplot_beforeandafterfilt.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/HDF-Data-QC"
  )

```

## Normalisation - Normalise gene expression per cell by the total gene expression within that cell. This is then multiplied by a default scale factor (10,000), and log-transformed.
```{r}
hdfseurat_filt <- NormalizeData(hdfseurat_filt)
```

### Session information
```{r session_info_99}
sessionInfo()
```

This document was processed on: `r Sys.Date()`.
