---
title: "7_Teo"
author: "BH"
date: "2024-03-15"
output:
 pdf_document:
    toc: yes
---
```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
```

```{r load_packages}
library("dplyr")
library("Seurat")
library("patchwork")
library("readxl")
library("tidyseurat")
library("ggplot2")
library("tidyverse")
library("RCurl")
library("cowplot")
library("sctransform")
library("scPred")
library("Seurat")
library("magrittr")
library("caret")
library("dplyr")
library("janitor")
library("factoextra")
library("reshape2")

options(future.globals.maxSize = 12000 * 1024^2)

```
# Investigating performance of EP/ES/DS SenPred 2D classifier on publicly available paper: Teo, Y.V., Rattanavirotkul, N., Olova, N., Salzano, A., Quintanilla, A., Tarrats, N., Kiourtis, C., MÃ¼ller, M., Green, A.R., Adams, P.D. and Acosta, J.C., 2019. Notch signaling mediates secondary senescence. Cell reports, 27(4), pp.997-1007.

This will allow the model to be tested on an alternative trigger: Oncogene Induced Senescence (OIS/RIS), and paracrine senescence (GFP).

## Loading data
```{r}

custom_theme <- readRDS(file = "custom_theme.rds")
HDF_Chan_combined_scPred <- readRDS(file ="HDF_Chan_combined_scPred.rds")

#install.packages("GEOquery")
library(GEOquery)
metadata <- read.csv("../Inputfiles/barcodes_Teo.csv")
counts <- read.table("../Inputfiles/GSE115301_Growing_Sen_10x_count.txt")

rownames(metadata) <- metadata$X
metadata = subset(metadata, select = -c(1))

#Creating Seurat object
teo <- CreateSeuratObject(counts = counts)   
teo_seurat <- AddMetaData(teo, metadata)

teo_seurat@meta.data <- merge(teo_seurat@meta.data, metadata,
                          by = 'row.names', all = TRUE)

teo_seurat@meta.data = subset (teo_seurat@meta.data, select = -c(6))
rownames(teo_seurat@meta.data) <- teo_seurat@meta.data$Row.names
teo_seurat@meta.data = subset (teo_seurat@meta.data, select = -c(1))

saveRDS(teo_seurat, file = "../Output/Seurat_Objects/Teo_seurat.RDS")
```


```{r}
Teo_seurat <- readRDS(file = "../Output/Seurat_Objects/Teo_seurat.RDS")

#Generating log10 genes per UMI for QC
Teo_seurat$log10GenesPerUMI <- log10(Teo_seurat$nFeature_RNA) / log10(Teo_seurat$nCount_RNA)

#Generating percentage mitochondrial features for QC
Teo_seurat[["percent.mt"]] <- PercentageFeatureSet(Teo_seurat, pattern = "^MT-")


#Visualise the number of cells per sample
Teo_numbers <- Teo_seurat@meta.data %>% 
  	ggplot(aes(x = Condition2.x, fill = Condition2.x)) + 
  	geom_bar() + 
  scale_fill_manual(values = c("#C22A3D", "darkgreen", "#2232A1"))+
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle("NCells") + custom_theme(base_size=20) +
  labs(x = "Condition", fill="Condition")


Teo_numbers

ggsave(
  "TeoNumbersPerSample.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/Teo_QC"
  )
```
## Number of UMIs per cell
```{r}
#Number of UMIs (transcripts) per cell
Teo_UMI <- Teo_seurat@meta.data %>% 
  	ggplot(aes(x=nCount_RNA, fill= Condition2.x)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 1000)+
  labs(fill="Condition")+
  custom_theme(base_size=20)

Teo_UMI

ggsave(
  "TeoNumberofUMI.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/Teo_QC"
  )
```

## Distribution of genes per cell
```{r}
# Visualize the distribution of genes detected per cell via histogram
Teo_features <- Teo_seurat@meta.data %>% 
  	ggplot(aes(x=nFeature_RNA, fill= Condition2.x)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 400)+
  labs(fill="Condition")+
  custom_theme(base_size=20)

Teo_features

ggsave(
  "Teofeaturespercell.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/Teo_QC"
  )
```

```{r}
# Visualize the distribution of genes detected per cell via boxplot
Boxplot_genespercell_Teo <- Teo_seurat@meta.data %>% 
  	ggplot(aes(x=Condition2.x, y=log10(nFeature_RNA), fill=Condition2.x)) + 
  	geom_boxplot() + 
  	theme_classic() +
    scale_fill_manual(values = c("#C22A3D", "darkgreen", "#2232A1"))+
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle("NCells vs NGenes")+
  custom_theme(base_size=20)+
  labs(fill="Condition", x="Condition")

Boxplot_genespercell_Teo

ggsave(
  "Teofeaturespercell_boxplot.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/Teo_QC"
  )
```

```{r}
Feature_count_mt <- Teo_seurat@meta.data %>% 
  	ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percent.mt)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 1000) +
  	geom_hline(yintercept = 400) +
  	facet_wrap(~Condition2.x)+
  custom_theme(base_size=10)

Feature_count_mt

ggsave(
  "Teofeature_count_mt.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/Teo_QC"
  )
```

## Mitochondrial gene expression distribution
```{r}
# Visualize the distribution of mitochondrial gene expression detected per cell
percentmt_Teo<- Teo_seurat@meta.data %>% 
  	ggplot(aes(x=percent.mt, fill=Condition2.x)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 12)+
  custom_theme(base_size=20)+
  labs(fill="Condition")

percentmt_Teo

ggsave(
  "Teo_mt.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/Teo_QC"
  )
```

## Log10 genes per UMI
```{r}
#Log10 genes per UMI
teo_log10umi <- Teo_seurat@meta.data %>%
  	ggplot(aes(x=log10GenesPerUMI, fill=Condition2.x)) +
  	geom_density(alpha = 0.2) +
  	theme_classic() +
  	geom_vline(xintercept = 0.8)+
  custom_theme(base_size=20)+
  labs(fill="Condition")

teo_log10umi

ggsave(
  "Teo_log10umi.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/Teo_QC"
  )

```
Now that we have visualized the various metrics, we can decide on the thresholds to apply which will result in the removal of low quality cells. 

To remove potentially apoptotic or 'low quality' cells we discarded cells with less than 400 expressed genes, and/or more than 12% mitochondrial reads. We also filtered out cells with a UMI count less than 1000. 

nCount_RNA > 1000
nFeature_RNA > 400
percent.mt <= 12

# Subsetting seurat object, and normalising
```{r}
Teo_seurat2 <- subset(x=Teo_seurat, 
                           subset= (percent.mt <=12) &
                            (nFeature_RNA >=400) &
                            (nCount_RNA >= 1000))


df.list <- list(Original=Teo_seurat@meta.data, Filtered=Teo_seurat2@meta.data)
dat <- stack(lapply(df.list, `[[`, "Condition2.x"))
ggplot(dat, aes(x=values, fill=ind)) + geom_bar(position="dodge")

#LogNormalize data
Teo_seurat2 <- NormalizeData(Teo_seurat2)
```
# Identifying highly variable features
```{r}
#Selecting highly variable features - high cell to cell variation
Teo_seurat2 <- FindVariableFeatures(Teo_seurat2, selection.method ="vst", nfeatures=2000)

#Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(Teo_seurat2), 10)
print(top10)

top2000 <- head(VariableFeatures(Teo_seurat2), 2000)
top2000

#Plot variable features without labels
plot1 <- VariableFeaturePlot(Teo_seurat2)
#Plot variable features with labels
plot2 <- LabelPoints(plot=plot1, points = top10, repel=TRUE)
plot1
plot2

```
##Scaling for dimensionality reduction

```{r}
#Scaling for dimensionality reduction
all.genes <- rownames(Teo_seurat2)
Teo_seurat2 <- ScaleData(Teo_seurat2, features = all.genes)

#Performing a PCA
Teo_seurat2 <- RunPCA(Teo_seurat2, features=VariableFeatures(object = Teo_seurat2))

PCA_Teo <- DimPlot(Teo_seurat2, reduction = "pca")
PCA_Teo

ggsave(
  "Teo_PCA.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/Teo_QC"
  )


```
```{r}
#Heatmaps of genes contributing to top PCs
DimHeatmap(Teo_seurat2, dims=1, cells=500, balanced=TRUE)
DimHeatmap(Teo_seurat2, dims=1:15, cells=500, balanced=TRUE)
ElbowPlot(Teo_seurat2)

#Use K-nearest neighbours based on PCA distance in FindNeighbours() function based on previously definited dimensions (10) from elbow plot
Teo_seurat2 <- FindNeighbors(Teo_seurat2, dims =1:10)

#group cells using findclusters function
Teo_seurat2 <- FindClusters(Teo_seurat2, resolution = 0.4)

#Look at cluster ID's of first 5 cells
head(Idents(Teo_seurat2), 5)
```
## Running UMAP
```{r}
#Running a UMAP
Teo_seurat2 <- RunUMAP(Teo_seurat2, dims=1:10)

plot <- DimPlot(Teo_seurat2, reduction = "umap", seed = 123) +
  custom_theme(base_size=20)
plot
ggsave(
  "UMAP_clusters_all.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/Teo"
  )
```

```{r}
plot2 <- DimPlot(Teo_seurat2, reduction = "umap",group.by = "Condition2.x", seed = 123)+
  custom_theme(base_size=20)
plot2
ggsave(
  "UMAP_clusters_samples.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/Teo"
  )
```

#Running SenPred 2D EP/ES/DS on Teo data
```{r}
query <- NormalizeData(Teo_seurat2)
query <- scPredict(query, HDF_Chan_combined_scPred)


DimPlot(query, group.by = "Condition2.x", label = FALSE, repel = TRUE, label.size=2)
pred_plot <- DimPlot(query, group.by = "scpred_prediction", label = FALSE, repel = TRUE, seed = 123) +
  custom_theme(base_size=20)

pred_plot


ggsave(
  "UMAP_clusters_scpred2d_EPESDS.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/Teo"
  )

```
##Extracting percentage predictions for Teo data
```{r}
query2 <- query@meta.data
table <- table(query2[, c("Condition2.x", "scpred_prediction")])
table
mat <- as.data.frame.matrix(table)
print(mat)

## Coverting numbers per condition into percentages
percentages <- (mat/rowSums(mat))*100
percentages
percentages <- na.omit(percentages)
percentages
as.matrix(percentages)
percentages <- percentages[ order(row.names(percentages)), ]
v <- rownames(percentages)

print(percentages)

df <- tibble::rownames_to_column(percentages, "Condition")
percentage2 <- melt(df)

##Reordering variables
percentage2$variable <- factor(percentage2$variable, levels = c("unassigned", "EP2D", "PDL50"))


percplot <- ggplot(percentage2, aes(x=Condition, y=value, fill=variable)) + 
    geom_bar(position='stack', stat='identity') +
  scale_fill_manual(values=c("antiquewhite4", "deeppink3", "#228c22")) +
  custom_theme(base_size=20)

percplot

ggsave(
  "barplot_scpred_percentages.tiff",
  plot = last_plot(),
  device = "tiff",
  path = "../Output/Teo"
  )

```
### Session information
```{r session_info_99}
sessionInfo()
```